File: Data Structures & Algorithms/eating-bananas/submission-1.cs
public class Solution {
    public int MinEatingSpeed(int[] piles, int h) {
        int l = 1, r = piles.Max();

        int res = r;
        while (l <= r) {
            int m = l + ((r - l) / 2);

            int times = 0;
            foreach (int p in piles) {
                times += (p + m - 1) / m;
            }

            if (times <= h) {
                res = Math.Min(res, m);
                r = m - 1;
            }  else {
                l = m + 1;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/insert-new-interval/submission-1.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
            if not intervals:
                return [newInterval]

            l, r = 0, len(intervals) - 1

            while l <= r:
                mid = (l + r) // 2

                if intervals[mid][0] < newInterval[0]:
                    l = mid + 1
                elif intervals[mid][0] > newInterval[0]:
                    r = mid - 1
                else:
                    l = mid
                    break
            intervals.insert(l, newInterval)

            res = []
            min_start = intervals[0][0]
            max_end = intervals[0][1]
            for start, end in intervals:
                if start > max_end:
                    res.append([min_start, max_end])
                    min_start = start
                    max_end = end
                else:
                    max_end = max(max_end, end)
                    min_start = min(min_start, start)
            res.append([min_start, max_end])
            return res
